genie-web/src/main/java/com/netflix/genie/web/agent/services/impl/AgentRoutingServiceCuratorDiscoveryImpl.java

```diff
@@ -0,0 +1,306 @@
+/*
+ *
+ *  Copyright 2020 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.agent.services.impl;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.netflix.genie.common.internal.util.ExponentialBackOffTrigger;
+import com.netflix.genie.common.internal.util.GenieHostInfo;
+import com.netflix.genie.web.agent.services.AgentRoutingService;
+import com.netflix.genie.web.util.MetricsUtils;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.listen.Listenable;
+import org.apache.curator.framework.state.ConnectionState;
+import org.apache.curator.framework.state.ConnectionStateListener;
+import org.apache.curator.x.discovery.ServiceDiscovery;
+import org.apache.curator.x.discovery.ServiceInstance;
+import org.apache.curator.x.discovery.ServiceType;
+import org.springframework.scheduling.TaskScheduler;
+
+import javax.validation.constraints.NotBlank;
+import java.time.Instant;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Implementation of {@link AgentRoutingService} that relies on Curator's Discovery extension.
+ * Rather than the traditional use of this recipe (register a service for the node itself, this class registers one
+ * service instance for each agent locally connected.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class AgentRoutingServiceCuratorDiscoveryImpl implements AgentRoutingService {
```

Comment:
It might be nice eventually for either this bean directly or maybe just a leader task to implement `InfoContributor` and dump a mapping of which job id's are connected to which nodes